Modules/hpe3par_snapshot.py

```diff
@@ -435,13 +463,190 @@ def restore_snapshot_online(
+
+    freq = "@hourly"
+    if retentionHours > expirationHours:
+       return (False, False, "Expiration time must be greater than or equal to retention time", {})   
+
+    if task_freq_custom and task_freq:
+       return (False, False, "Provide either task_freq or task_freq_custom", {})        
+
+    if task_freq_custom:          
+       if ' ' in task_freq_custom:            
+           task_custom_list = str(task_freq_custom).split()
+           if len(task_custom_list) == 5:
+               minutes_task = task_custom_list[0]               
+               hour_task = task_custom_list[1]
+               day_task = task_custom_list[2]
+               month_task = task_custom_list[3]  
+               day_of_week_task = task_custom_list[4]
+               if '*' not in minutes_task:
+                  if (int(minutes_task) > 50 or int(minutes_task) < 0):
+                    return (False, False, "Invalid task frequency minutes should be between 0-50", {})
+               else:
+                  if (len(minutes_task)>1 or minutes_task == ""):
+                    return (False, False, "Invalid task frequency minutes", {})
+               if '*' not in hour_task:
+                  if (int(hour_task) > 23 or int(hour_task) < 0):
+                    return (False, False, "Invalid task frequency hours should be between 0-23", {})
+               else:
+                  if (len(hour_task)>1 or hour_task == ""):
+                    return (False, False, "Invalid task frequency hours", {})
+               if '*' not in day_task:
+                  if (int(day_task) > 31 or int(day_task) < 1):
+                    return (False, False, "Invalid task frequency day should be between 1-31", {})
+               else:
+                  if (len(day_task)>1 or day_task == ""):
+                    return (False, False, "Invalid task frequency day", {})
+               if '*' not in month_task:
+                  if (int(month_task) > 12 or int(month_task) < 1):
+                    return (False, False, "Invalid task frequency month should be between 1-12", {})
+               else:
+                  if (len(month_task)>1 or month_task == ""):
+                    return (False, False, "Invalid task frequency month", {})
+               if '*' not in day_of_week_task:
+                  if (int(day_of_week_task) > 6 or int(day_of_week_task) < 0):
+                    return (False, False, "Invalid task frequency day of week should be between 0-6", {})
+               else:
+                  if (len(day_of_week_task)>1 or day_of_week_task == ""):
+                    return (False, False, "Invalid task frequency day of week", {})
+           else:
+              return (False, False, "Invalid task frequency string", {})
+       else:
+             return (False, False, "Invalid task frequency string", {})
+       freq = task_freq_custom        
+    try:
+        client_obj.login(storage_system_username, storage_system_password)
```

Comment:
Did we check if we really need to login?

Reply:
yes, we need wsapi method here for volumeExists